### PR TITLE
add checking for rgw frontend init

### PR DESCRIPTION
**Pull request by qiuxinyidian**
fixes: https://tracker.ceph.com/issues/63644

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_lib.cc [hunk 1]
```diff
@@ -524,7 +525,12 @@ namespace rgw {
     register_async_signal_handler(SIGUSR1, rgw::signal::handle_sigterm);
 
     main.init_tracepoints();
-    main.init_frontends2(this /* rgwlib */);
+    r = main.init_frontends2(this /* rgwlib */);
+    if (r != 0) {
+      derr << "ERROR: unable to initialize frontend, r = " << r << dendl;
+      exit(1);
```
Comment (cbodley) [inline src/rgw/rgw_lib.cc:531]: RGWLib is a library interface that allows other processes like nfs-ganesha to load rgw, so calling `exit()` is not the right way to handle errors. this error code should be returned by `RGWLib::init()`, though there's probably some cleanup we'd need to do first

Comment (qiuxinyidian) [reply src/rgw/rgw_lib.cc:531]: @cbodley  changing exit(1) to return -r ?

Comment (cbodley) [reply src/rgw/rgw_lib.cc:531]: `return r`, but call `main.shutdown()` first to clean up what we've initialized so far

**Review by cbodley — APPROVED** [review]
manually tested the rgw_main.cc changes using `--rgw-frontends='beast port=80'` as a non-root user:
```
2023-11-30T14:18:44.279-0500 9081180 -1 failed to bind address 0.0.0.0:80: Permission denied
2023-11-30T14:18:44.288-0500 9081180 -1 ERROR: failed initializing frontend 
2023-11-30T14:18:44.289-0500 9081180 -1 ERROR:  initialize frontend fail, r = 13
```
radosgw exits cleanly after

**Review by mattbenjamin — APPROVED** [review]
maybe squash these, but looks fine

**Comment by qiuxinyidian** [comment]
@cbodley  why this PR not close after approved? is there any thing wrong ?

**Comment by cbodley** [comment]
> @cbodley why this PR not close after approved? is there any thing wrong ?

approval means a successful code review. the needs-qa tag means we still need to run it through regression tests before merging

in the meantime, could you please squash your 3 commits into a single one?

**Pull request merged**